src/gallia/services/uds/ecu.py

```diff
@@ -349,92 +338,100 @@ async def transmit_data(
-        self.tester_present_task = asyncio.create_task(coroutine)
-        self.tester_present_task.add_done_callback(
-            handle_task_error,
-            context=set_task_handler_ctx_variable(__name__, "TesterPresent"),
-        )
 
-        # enforce context switch
-        # this ensures, that the task is executed at least once
-        # if the task is not executed, task.cancel will fail with CancelledError
-        await asyncio.sleep(0)
+    async def _send_tester_present(self, timeout: float = 1) -> bool:
+        """
+        Send TesterPresent to ECU and return `True` on any UDS response.
+        """
+
+        logger.debug("Sending TesterPresent request to ECU...")
+        try:
+            await self.tester_present(
+                config=UDSRequestConfig(
+                    max_retry=0,
+                    tags=["tp"],
+                    timeout=timeout,
+                    skip_hooks=True,
+                )
+            )
+
+        # When the ECU is not ready, we expect an UDSException, e.g. MissingResponse.
+        # On ConnectionError, we additionally reconnect the transport to ensure connectivity.
+        # Since Gallia converts a ConnectionError to a MissingResponse in `request_unsafe`, however,
+        # there is a need to reconnect the transport also in case of high-level UDSExceptions
+        # such as MissingResponses that are raised (__cause__) from ConnectionErrors.
+        except (ConnectionError, UDSException) as e:
+            logger.debug(f"TesterPresent failed: {e!r}")
+            if isinstance(e, ConnectionError) or isinstance(e.__cause__, ConnectionError):
+                logger.info("Connection lost; reconnecting...")
+                await self.reconnect()
+            return False
+
+        except Exception as e:
+            logger.warning(f"TesterPresent got {e!r}")
+            return False
+
+        return True
+
+    async def attach_tester_present_sender(
+        self,
+        interval: float,
+        ignore_activity: bool = False,
+        task_timeout: float = 1,
+    ) -> None:
+        logger.debug("Attaching TesterPresentSender")
+
+        if self.tester_present_task is None:
+            self.tester_present_task = TesterPresentSender(
+                interval,
+                lambda: self._send_tester_present(timeout=task_timeout),
+                oneshot=False,
+                ignore_activity=ignore_activity,
+            )
+
+        await self.tester_present_task.start()
+
+    async def detach_tester_present_sender(self) -> None:
+        logger.debug("Detaching TesterPresentSender")
 
-    async def stop_cyclic_tester_present(self) -> None:
-        logger.debug("Stopping tester present worker")
         if self.tester_present_task is None:
-            logger.warning("BUG: stop_cyclic_tester_present() called but no task running")
+            logger.warning("Attempted to detach TesterPresentSender, but it is already gone!")
```

Comment:
Same here, isn't this a programming error? Maybe raise AssertionError (cf. below)?